### PR TITLE
Make TM work with --recover

### PR DESCRIPTION
**Pull request by permcody**
refs #7548

@sapitts - This fixes the recover problem with TensorMechanics. I also did some casual optimizations to some of these materials.

### CODE REVIEW

File: modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C [hunk 1]
```diff
@@ -27,19 +27,13 @@ StressDivergenceRSphericalTensors::StressDivergenceRSphericalTensors(const Input
 Real
 StressDivergenceRSphericalTensors::computeQpResidual()
 {
-  Real div = 0;
+  mooseAssert(_assembly.coordSystem() == Moose::COORD_RSPHERICAL,
```
Comment (friedmud) [inline modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C]: Is this really necessary to do in `computeQpResidual()`?  Do we possibly have a `subdomainSetup()` or something we can use instead?  In `dbg` mode this is a bummer (but not a HUGE) bummer.

Comment (permcody) [inline modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C]: We can do it in intialSetup().  I was just avoiding changing too much code but this change is worth it. I'll update and repush...

File: modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C [hunk 2]
```diff
@@ -70,6 +64,8 @@ StressDivergenceRSphericalTensors::calculateJacobian(unsigned int ivar, unsigned
 {
   RealGradient test_r, phi_r;
 
+  mooseAssert(ivar == 0 && jvar == 0, "Invalid component in Jacobian Calculation"); // Only nonzero case for a 1D simulation
```
Comment (friedmud) [inline modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C]: Wait - what the hell is `0` doing here?  That's going to be wrong unless `disp_r` is the first variable in the system!!!

The variable number for `disp_r` needs to be saved off from `coupled()` and compared to!

File: modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C [hunk 3]
```diff
@@ -70,6 +64,8 @@ StressDivergenceRSphericalTensors::calculateJacobian(unsigned int ivar, unsigned
...
+
   if (ivar == 0)  //Case grad_test for r, requires contributions from stress_{rr}, stress_{\theta \theta}, and stress_{\phi \phi}
```
Comment (friedmud) [inline modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C]: Same here - this is insane!

File: modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C (hunk 2 above)
Comment (friedmud) [inline modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C]: Wait - now I see... ugh.

File: modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C (hunk 3 above)
Comment (friedmud) [inline modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C]: Nvm

File: modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C (hunk 2 above)
Comment (permcody) [inline modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C]: @sapitts, perhaps you can shed some light on this. I wasn't really reviewing the code for correctness. I was just making it work.

**Comment by friedmud** [comment]
It's outside the scope of this PR... but I wonder if we need a `CoordinateSystemInterface` where you can call `requiresRZ()`, etc. and then `FEProblem::checkProblemIntegrity()` can make sure that any objects that require a specific type of coordinate system are only defined on blocks of that type...

File: modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C (hunk 2 above)
Comment (friedmud) [inline modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C]: I see now - they are basically switching to a "local" indexing in `computeQpJacobian()`

File: modules/tensor_mechanics/src/materials/ComputeAxisymmetricRZFiniteStrain.C [hunk 4]
```diff
@@ -25,8 +27,10 @@ ComputeAxisymmetricRZFiniteStrain::ComputeAxisymmetricRZFiniteStrain(const Input
 void
 ComputeAxisymmetricRZFiniteStrain::initialSetup()
 {
-  if (_assembly.coordSystem() != Moose::COORD_RZ)
-    mooseError("The coordinate system must be set to RZ for Axisymmetric simulations.");
+  const auto & subdomainIDs = _mesh.meshSubdomains();
```
Comment (friedmud) [inline modules/tensor_mechanics/src/materials/ComputeAxisymmetricRZFiniteStrain.C]: This is a better solution than the above ones in `computeQpResidual()`... can we do this there too?

**Comment by permcody** [comment]
Yeah, that's not a bad idea. Why don't you open a ticket that we'll never get to... 😄

**Comment by friedmud** [comment]
What was the actual recovery problem?

File: modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C (hunk 2 above)
Comment (sapitts) [inline modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C]: This 0 and the one marked below would be my fault -- a direct copy from Solid Mechanics.  I can try to tackle this in a few weeks?

**Comment by permcody** [comment]
They are inspecting the result of _assembly.coordSystem() which is there for getting the reference to the current CoordSystem on each element. I guess we are getting lucky that this "works" on initialSetup() now. It's not really well-defined to expect that to equal anything useful at that stage. Even if it did, it wouldn't be correct for a mixed coordinate type simulation.

File: modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C (hunk 2 above)
Comment (dschwen) [inline modules/tensor_mechanics/src/kernels/StressDivergenceRSphericalTensors.C]: Probably shouldn't be called `ivar` and `jvar` but `icomponent` and `jcomponent` or something like that.

**Pull request merged**